DeepSpeech.ipynb

```diff
@@ -1320,170 +1292,390 @@
-    "            training_time = stopwatch(training_time)\n",
-    "            \n",
-    "            # Add batch to loss\n",
-    "            loss += result[1]\n",
+    "        is_validation_step = (epoch > 0 and (epoch + 1) % validation_step == 0) or epoch == training_iters - 1\n",
     "\n",
-    "            # Log all variable states in current step\n",
-    "            step = epoch * total_batches + batch * len(available_devices)\n",
-    "            writer.add_summary(result[2], step)\n",
-    "            if do_fulltrace:\n",
-    "                writer.add_run_metadata(loss_run_metadata, 'loss_epoch%d_batch%d'   % (epoch, batch))\n",
-    "            writer.flush()\n",
-    "            \n",
-    "            # Collect individual sample results for WER report\n",
-    "            if is_display_step:\n",
-    "                collect_results(train_results, result[3])\n",
-    "                # Add batch to total_accuracy\n",
-    "                total_accuracy += result[4]\n",
-    "        \n",
-    "        # Print WER report\n",
-    "        if is_display_step:\n",
-    "            train_wer = calculate_and_print_wer_report( \\\n",
-    "                \"Training avg_cer=\" + \"{:.9f}\".format((total_accuracy / ceil(batches_per_device))), train_results)\n",
-    "            print\n",
+    "        print \"Training model...\"\n",
+    "        train_time = stopwatch(train_time)\n",
+    "        result = apply_batches(train_context, session, epoch=epoch, query_report=is_display_step)\n",
+    "        train_time = stopwatch(train_time)\n",
+    "\n",
+    "        result = print_report(\"Training\", result)\n",
+    "        # If there was a WER calculated, we keep it\n",
+    "        train_wer = result if result is not None else train_wer\n",
     "            \n",
     "        # Validation step\n",
     "        if is_validation_step:\n",
-    "            dev_wer = run_inference(session, \"Validation\", data_sets.dev)\n",
-    "\n",
-    "        # Checkpoint the model\n",
-    "        if is_checkpoint_step:\n",
+    "            print \"Hibernating training session into directory\", \"%s\" % checkpoint_dir\n",
     "            checkpoint_path = os.path.join(checkpoint_dir, 'model.ckpt')\n",
-    "            print \"Checkpointing in directory\", \"%s\" % checkpoint_dir\n",
-    "            saver.save(session, checkpoint_path, global_step=epoch)\n",
-    "            print\n",
+    "            # If the hibernation_path is set, the next epoch loop will load the model\n",
+    "            hibernation_path = stop_execution_context( \\\n",
+    "                train_context, session, checkpoint_path=checkpoint_path, global_step=epoch)\n",
     "            \n",
+    "            # Validating the moel in a fresh session\n",
+    "            print \"Validating model...\"\n",
+    "            result = run_set('dev', model_path=hibernation_path, query_report=(epoch == training_iters - 1))\n",
```

Comment:
Having this only on the last epoch seems counter to the purpose of printing out a validation WER. It should be configurable or maybe happen all the time but have `is_validation_step` set to a computationally reasonable value.

Traditionally one prints out a validation WER to monitor the generalization of the model as it trains. To do this one needs to see the training WER and the validation WER throughout training to see if the training WER dips significantly below the validation WER.

When this happens you want to stop training, as this is a signal of over fitting and subsequent epochs will only server to further overfit and thus waste GPU time.